mpv/0002-avcodec-h264-enable-libyami-h264-decoder.patch

```diff
@@ -0,0 +1,443 @@
++        return -1;
++    }
++
++    if((avctx->get_buffer2 != avcodec_default_get_buffer2) &&(avctx->get_buffer2 != NULL)) {
++        native_display.type = NATIVE_DISPLAY_VA;
++        native_display.handle = (intptr_t)(vactx->display);
++        s->decoder ->setNativeDisplay(&native_display);
++    }
++
++    memset(&config_buffer,0,sizeof(VideoConfigBuffer));
++    if (avctx->extradata && avctx->extradata_size && avctx->extradata[0] == 1) {
++        config_buffer.data = avctx->extradata;
++        config_buffer.size = avctx->extradata_size;
++    }
++    config_buffer.profile = VAProfileNone;
++    status = s->decoder->start(&config_buffer);
++    if (status != DECODE_SUCCESS) {
++        av_log(avctx, AV_LOG_ERROR, "yami h264 decoder fail to start\n");
++        return -1;
++    }
++
++    s->decode_count = 0;
++    s->decode_count_yami = 0;
++    s->render_count = 0;
++
++    avctx->get_format(avctx, &choices);
++
++    return 0;
++}
++
++static void* decodeBuffer(void *arg, VideoDecodeBuffer *in_buffer)
++{
++    AVCodecContext *avctx = (AVCodecContext*)arg;
++    YamiContext *s = (YamiContext*)avctx->priv_data;
++    Decode_Status status;
++
++    // decode one input buffer
++    if (in_buffer->data) {
++        status = s->decoder->decode(in_buffer);
++    }
++
++    if (DECODE_FORMAT_CHANGE == status) {
++        s->format_info = s->decoder->getFormatInfo();
++        // resend the buffer
++        status = s->decoder->decode(in_buffer);
++        avctx->width = s->format_info->width;
++        avctx->height = s->format_info->height;
++        avctx->pix_fmt = AV_PIX_FMT_YUV420P;
++    }
++
++    if (in_buffer->data)
++        s->decode_count_yami++;
++    av_free(in_buffer);
++
++    return NULL;
++}
++
++static int yami_decode_frame(AVCodecContext *avctx, void *data /* output frame */,
++                                    int *got_frame, AVPacket *avpkt /* input compressed data*/)
++{
++    YamiStatus ret;
++    AVFrame get_frame;
++    YamiContext *s = (YamiContext*)avctx->priv_data;
++    const VideoRenderBuffer* out_frame;
++    VideoDecodeBuffer *in_buffer = NULL;
++
++    if ((!s->vpp_init) &&(avctx->get_buffer2 != avcodec_default_get_buffer2) &&(avctx->get_buffer2 != NULL)) {
++        initDisplay(avctx);
++        s->vpp.reset(createVideoPostProcess(YAMI_VPP_SCALER), releaseVideoPostProcess);
++        if (!(s->vpp)) {
++            return -1;
++        }
++        if (s->vpp->setNativeDisplay(*(s->nativeDisplay)) != YAMI_SUCCESS) {
++            return -1;
++        }
++        s->vpp_init = true;
++    }
++
++    // append avpkt to input buffer queue
++    in_buffer = (VideoDecodeBuffer*)av_mallocz(sizeof(VideoDecodeBuffer));
++    in_buffer->data = avpkt->data;
++    in_buffer->size = avpkt->size;
++    in_buffer->timeStamp = avpkt->pts;
++
++    decodeBuffer(avctx, in_buffer);
++    // get an output buffer from yami
++    if (avpkt->data)
++        out_frame = s->decoder->getOutput(false);
```

Comment:
better use SharedPtr<VideoFrame> getOutput(), since we will drop VideoRenderBuffer